File: apps/ai-service/routers/roadmap.py
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional
from services.llm import get_llm_response

router = APIRouter()

ROADMAP_CATEGORIES = [
    "DSA",
    "APTITUDE",
    "SYSTEM_DESIGN",
    "CORE_SUBJECTS",
    "PROJECTS",
    "RESUME_BUILDING",
]


class RoadmapGenerateRequest(BaseModel):
    branch: str
    graduation_year: int = 2026
    skill_level: str = "INTERMEDIATE"
    target_companies: list[str] = []
    study_hours_per_day: int = 3
    college: Optional[str] = None
    skills: Optional[list[str]] = None


def _fallback_roadmap(req: RoadmapGenerateRequest) -> dict:
    hours = req.study_hours_per_day
    companies = req.target_companies or ["Google", "Microsoft", "Amazon", "Flipkart"]
    daily_tasks = []
    categories_cycle = [
        ("DSA", "Solve 2 array/string problems on LeetCode"),
        ("APTITUDE", "Practice 20 quantitative aptitude questions"),
        ("DSA", "Revise trees/graphs — 1 medium problem"),
        ("CORE_SUBJECTS", "Study OS/DBMS concepts for 1 hour"),
        ("APTITUDE", "Logical reasoning mock set (30 min)"),
        ("PROJECTS", "Work on portfolio project feature"),
        ("DSA", "Dynamic programming pattern practice"),
    ]
    for day in range(1, 15):
        cat, title = categories_cycle[(day - 1) % len(categories_cycle)]
        daily_tasks.append(
            {
                "id": f"day-{day}",
                "day": day,
                "category": cat,
                "title": title,
                "duration_minutes": min(hours * 60, 180),
                "description": f"Day {day} focus for {req.branch} placement prep.",
            }
        )

    return {
        "summary": f"Personalized {req.branch} placement roadmap targeting {', '.join(companies[:3])}.",
        "timeline_months": 6,
        "adaptive_tips": [
            "Increase DSA hours if mock test scores stay below 70%.",
            "Add company-specific aptitude sets 8 weeks before interviews.",
            "Schedule weekly mock interviews from month 4 onward.",
        ],
        "categories": {
            cat: {
                "priority": "high" if cat in ("DSA", "APTITUDE") else "medium",
                "weekly_hours": hours if cat == "DSA" else max(2, hours // 2),
                "focus_topics": _category_topics(cat, req.branch),
            }
            for cat in ROADMAP_CATEGORIES
        },
        "monthly_goals": [
            {
                "month": 1,
                "title": "Foundation — DSA & Aptitude basics",
                "categories": ["DSA", "APTITUDE", "CORE_SUBJECTS"],
                "targets": ["Complete 40 easy DSA problems", "Aptitude accuracy > 60%"],
            },
            {
                "month": 2,
                "title": "Intermediate patterns & core subjects",
                "categories": ["DSA", "CORE_SUBJECTS", "APTITUDE"],
                "targets": ["30 medium DSA problems", "Finish OS & DBMS revision"],
            },
            {
                "month": 3,
                "title": "Projects & resume",
                "categories": ["PROJECTS", "RESUME_BUILDING", "DSA"],
                "targets": ["Ship 1 portfolio project", "ATS resume score > 75"],
            },
            {
                "month": 4,
                "title": "System design & company prep",
                "categories": ["SYSTEM_DESIGN", "APTITUDE", "DSA"],
                "targets": ["5 system design case studies", "Company-specific aptitude"],
            },
            {
                "month": 5,
                "title": "Mock interviews & weak-area drills",
                "categories": ["DSA", "APTITUDE", "RESUME_BUILDING"],
                "targets": ["4 mock interviews", "Fix top 3 skill gaps"],
            },
            {
                "month": 6,
                "title": "Final revision & applications",
                "categories": ["DSA", "APTITUDE", "RESUME_BUILDING"],
                "targets": ["Apply to target companies", "Daily revision loops"],
            },
        ],
        "weekly_milestones": [
            {
                "week": w,
                "title": f"Week {w}: {_week_focus(w)}",
                "categories": ["DSA", "APTITUDE"] if w % 2 else ["DSA", "CORE_SUBJECTS"],
                "hours": hours * 7,
                "deliverable": f"Week {w} checkpoint — review progress and adjust plan.",
            }
            for w in range(1, 9)
        ],
        "daily_tasks": daily_tasks,
        "target_companies": companies,
        "skill_gaps": ["Dynamic Programming", "System Design", "Aptitude speed"],
    }


def _category_topics(category: str, branch: str) -> list[str]:
    topics = {
        "DSA": ["Arrays", "Strings", "Trees", "Graphs", "DP"],
        "APTITUDE": ["Quantitative", "Logical reasoning", "Data interpretation"],
        "SYSTEM_DESIGN": ["Scalability", "Caching", "Load balancing", "Databases"],
        "CORE_SUBJECTS": ["OS", "DBMS", "Networks", "OOP"],
        "PROJECTS": ["Full-stack app", "Open source contribution"],
        "RESUME_BUILDING": ["ATS keywords", "Impact metrics", "GitHub profile"],
    }
    base = topics.get(category, [])
    if branch.lower().startswith("cse") or "computer" in branch.lower():
        return base
    return base[:3]


def _week_focus(week: int) -> str:
    focuses = [
        "Arrays & Strings mastery",
        "Linked lists, stacks & queues",
        "Trees & graph traversals",
        "Dynamic programming patterns",
        "Core CS subjects revision",
        "System design fundamentals",
        "Projects & resume polish",
        "Mock tests & interview prep",
    ]
    return focuses[min(week - 1, len(focuses) - 1)]


@router.post("/generate")
async def generate_roadmap(req: RoadmapGenerateRequest):
    companies = ", ".join(req.target_companies) if req.target_companies else "top product companies"
    prompt = f"""Create a detailed placement preparation roadmap as JSON for an Indian college student.

Profile:
- Branch: {req.branch}
- Graduation year: {req.graduation_year}
- Current skill level: {req.skill_level}
- Target companies: {companies}
- Study hours per day: {req.study_hours_per_day}
- College: {req.college or "N/A"}
- Existing skills: {req.skills or []}

Return ONLY valid JSON with this exact structure:
{{
  "summary": "2-3 sentence overview",
  "timeline_months": 6,
  "adaptive_tips": ["tip for adjusting plan based on progress"],
  "categories": {{
    "DSA": {{ "priority": "high|medium|low", "weekly_hours": number, "focus_topics": ["topic"] }},
    "APTITUDE": {{ ... }},
    "SYSTEM_DESIGN": {{ ... }},
    "CORE_SUBJECTS": {{ ... }},
    "PROJECTS": {{ ... }},
    "RESUME_BUILDING": {{ ... }}
  }},
  "monthly_goals": [
    {{ "month": 1, "title": "...", "categories": ["DSA"], "targets": ["measurable target"] }}
  ],
  "weekly_milestones": [
    {{ "week": 1, "title": "...", "categories": ["DSA"], "hours": number, "deliverable": "..." }}
  ],
  "daily_tasks": [
    {{
      "id": "day-1",
      "day": 1,
      "category": "DSA|APTITUDE|SYSTEM_DESIGN|CORE_SUBJECTS|PROJECTS|RESUME_BUILDING",
      "title": "specific task",
      "duration_minutes": number,
      "description": "actionable steps"
    }}
  ],
  "target_companies": ["company names"],
  "skill_gaps": ["gap to address"]
}}

Requirements:
- Include at least 14 daily_tasks spanning all 6 categories
- Include 6 monthly_goals and 8 weekly_milestones
- Scale difficulty to skill level ({req.skill_level})
- Personalize for {req.branch} branch and {companies}
- daily_tasks ids must be unique strings like day-1, day-2, etc."""

    try:
        result = await get_llm_response(prompt, json_mode=True)
        if not result.get("daily_tasks"):
            return _fallback_roadmap(req)
        return result
    except Exception:
        return _fallback_roadmap(req)

File: apps/api/src/demo/roadmap.ts
export const demoRoadmapPlan = {
  summary:
    "Demo placement roadmap — explore DSA, aptitude, system design, and interview prep on a 6-month timeline.",
  timeline_months: 6,
  adaptive_tips: [
    "Complete daily tasks consistently to unlock adaptive plan adjustments.",
    "Focus extra hours on your weakest category each week.",
  ],
  categories: {
    DSA: { priority: "high", weekly_hours: 12, focus_topics: ["Arrays", "Trees", "DP"] },
    APTITUDE: { priority: "high", weekly_hours: 6, focus_topics: ["Quant", "Logical"] },
    SYSTEM_DESIGN: { priority: "medium", weekly_hours: 4, focus_topics: ["Scalability", "Caching"] },
    CORE_SUBJECTS: { priority: "medium", weekly_hours: 5, focus_topics: ["OS", "DBMS"] },
    PROJECTS: { priority: "medium", weekly_hours: 4, focus_topics: ["Portfolio app"] },
    RESUME_BUILDING: { priority: "medium", weekly_hours: 2, focus_topics: ["ATS", "Metrics"] },
  },
  monthly_goals: [
    { month: 1, title: "DSA foundations", categories: ["DSA", "APTITUDE"], targets: ["40 easy problems"] },
    { month: 2, title: "Core subjects", categories: ["CORE_SUBJECTS", "DSA"], targets: ["OS/DBMS revision"] },
  ],
  weekly_milestones: [
    { week: 1, title: "Arrays & Strings", categories: ["DSA"], hours: 21, deliverable: "10 problems solved" },
    { week: 2, title: "Trees & Graphs", categories: ["DSA"], hours: 21, deliverable: "8 medium problems" },
  ],
  daily_tasks: [
    {
      id: "day-1",
      day: 1,
      category: "DSA",
      title: "Arrays — two-pointer patterns",
      duration_minutes: 90,
      description: "Solve 2 easy + 1 medium array problems.",
    },
    {
      id: "day-2",
      day: 2,
      category: "APTITUDE",
      title: "Quantitative aptitude drill",
      duration_minutes: 60,
      description: "20 questions timed — review mistakes.",
    },
    {
      id: "day-3",
      day: 3,
      category: "CORE_SUBJECTS",
      title: "OS — processes & threads",
      duration_minutes: 60,
      description: "Read notes + 10 conceptual MCQs.",
    },
  ],
  target_companies: ["Google", "Microsoft", "Amazon"],
  skill_gaps: ["Dynamic Programming", "System Design"],
};

export const demoRoadmapRecord = {
  id: "demo-roadmap",
  userId: "demo-guest",
  branch: "Computer Science",
  graduationYear: 2026,
  skillLevel: "INTERMEDIATE",
  targetCompanies: ["Google", "Microsoft", "Amazon"],
  studyHoursPerDay: 3,
  plan: demoRoadmapPlan,
  progressPercent: 33.3,
  adaptiveNotes: null as string | null,
  createdAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  completedTasks: [] as { taskKey: string; category: string | null; completedAt: string }[],
};

demoRoadmapRecord.completedTasks.push({
  taskKey: "day-1",
  category: "DSA",
  completedAt: new Date().toISOString(),
});
